Support specifying explicit SDK version

ojhunt opened the pull request:
#### 9140fbde2d9227d8df40b4db4daf9d1de71ad02f
<pre>
Support specifying explicit SDK version
<a href="https://bugs.webkit.org/show_bug.cgi?id=274298">https://bugs.webkit.org/show_bug.cgi?id=274298</a>

Reviewed by Elliott Williams.

Strip the version number from the specified SDK as well when setting
the platform name

* Tools/Scripts/webkitdirs.pm:
(determineXcodeSDKPlatformName):

Canonical link: <a href="https://commits.webkit.org/279009@main">https://commits.webkit.org/279009@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6163619899decf389b90da82807bdea515ad68db

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52228 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/31559 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4648 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/55502 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2951 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/54532 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37962 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2649 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/42507 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/1897 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54324 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/29190 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45078 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23578 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/26458 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/2352 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1110 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/45577 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/48363 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/2500 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57098 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/51736 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27354 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/2557 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/49897 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/28587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45194 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/49139 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11418 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/29499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/64043 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28331 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12124 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -917,6 +917,8 @@ sub availableXcodeSDKs
 
 sub isValidXcodeSDKPlatformName($) {
     my $name = shift;
+    # strip sdk version and variant qualifiers prior to validating platform
```
This comment says what the code does, but not why. I do not immediately understand why - `make` always works with the active Xcode anyway, so it seems like there would be no need to specify SDK version, as there's only one.

ojhunt replied:
@aproskuryakov if you have multiple sdk versions installed you cannot currently use the sdk argument to specify which sdk to use, e.g. `build-webkit --sdk macosx14.1`  will fail because "macosx14.1" is not in the list of valid platforms.

emw-apple commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -917,6 +917,8 @@ sub availableXcodeSDKs
 sub isValidXcodeSDKPlatformName($) {
     my $name = shift;
+    # strip sdk version and variant qualifiers prior to validating platform
+    $name =~ s/(\d+\.[\d\.]+)?(\.internal)?$//;
```
I don't think stripping the version in the validation function makes sense, because then we'd have to make sure that the rest of `webkitdirs` can handle when $xcodeSDKPlatformName contains a version number. What I would recommend instead is a modification to `determineXcodeSDKPlatformName`:

```diff
     if (checkForArgumentAndRemoveFromARGVGettingValue("--sdk", \$sdk)) {
         $xcodeSDK = lc $sdk;
         $xcodeSDKPlatformName ||= $sdk;
-        $xcodeSDKPlatformName =~ s/\.internal$//;
+        $xcodeSDKPlatformName =~ s/(\d+\.[\d\.]+)?(\.internal)?$//;
         die "Couldn't determine platform name from Xcode SDK" unless  isValidXcodeSDKPlatformName($xcodeSDKPlatformName);
         return;
     }
```

ojhunt replied:
that seems more correct to me assuming that's what gets passed to the validation code, I will try to verify in a bit

emw-apple reviewed (commented):
I agree with @aproskuryakov, I'm not entirely sure what you're trying to accomplish. Do you have a special Xcode configuration where you have multiple SDKs for the same platform, differing only in version?

emw-apple commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -965,8 +965,10 @@ sub determineXcodeSDKPlatformName {
+        $xcodeSDKPlatformName =~ s/(\d+\.[\d\.]+)?(\.internal)?$//;
         die "Couldn't determine platform name from Xcode SDK" unless isValidXcodeSDKPlatformName($xcodeSDKPlatformName);
+        `xcrun --sdk $xcodeSDK --show-sdk-path 2>&1`;
+        die "Unable to find requested sdk '" . $xcodeSDK . "' requested" if exitStatus($?);
```
Talked privately but: I've tried to avoid making calls to `xcrun` when executing `XcodeOptions()`, since xcrun can be a little slow (~2 sec) when uncached. So, I'd suggest to not unconditionally run this, especially since most people pass an unversioned SDK name.

emw-apple reviewed (approved):
r=me

ojhunt committed:
Support specifying explicit SDK version
https://bugs.webkit.org/show_bug.cgi?id=274298

Reviewed by Elliott Williams.

Strip the version number from the specified SDK as well when setting
the platform name

* Tools/Scripts/webkitdirs.pm:
(determineXcodeSDKPlatformName):

Canonical link: https://commits.webkit.org/279009@main
